### PR TITLE
[server] Fallback to unmanaged parameters

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

Follow up: https://lists.osgeo.org/pipermail/qgis-developer/2019-April/056979.html

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by pblottiere** [comment]
ping @dmarteau

File: src/server/qgsserverparameters.cpp [hunk 1]
```diff
@@ -531,7 +531,14 @@ QString QgsServerParameters::request() const
 
 QString QgsServerParameters::value( const QString &key ) const
 {
-  return value( QgsServerParameter::name( key ) ).toString();
+  if ( ! mParameters.contains( QgsServerParameter::name( key ) ) )
+  {
+    return mUnmanagedParameters[key];
```
Comment (elpaso) [inline src/server/qgsserverparameters.cpp]: what happens if `key` is not in `mUnmanagedParameters` ?

File: tests/src/python/test_qgsserver_request.py [hunk 2]
```diff
@@ -196,6 +196,11 @@ def _check_links(params, method='GET'):
         _check_links(params)
         _check_links(params, 'POST')
 
+    def test_add_parameters(self):
+        request = QgsServerRequest()
+        request.setParameter('FOOBAR', 'foobar')
+        self.assertEqual(request.parameter('FOOBAR'), 'foobar')
+
```
Comment (elpaso) [inline tests/src/python/test_qgsserver_request.py]:
```suggestion
self.assertEqual(request.parameter('DOESNOTEXIST'), '')  # Or assertIsNone?
```

File: src/server/qgsserverparameters.cpp (hunk 1 above)
Comment (pblottiere) [reply src/server/qgsserverparameters.cpp]: An empty string should be returned by qmap in this case. I added a unit test accordingly.

**Comment by dmarteau** [comment]
@pblottiere yes, is this what I had in mind.

Just as a suggestion for later improvement: I have noticed that at https://github.com/qgis/QGIS/blob/master/src/server/services/wms/qgswms.cpp#L73, the url is parsed to create the `QgsWmsParameters`; that means that we have the following workflow:

Parse url in QgisServerRequest -> Rebuild the url internally when setting parameters -> Parse url once again. 

Since we have all parameters available internally, may be it could be interesting to have a constructor method for building directly a QgsWmsParameters frow another QgsServerParameters  ?

i.e
```
const QgsWmsParameters parameters( request.serverParameters() );
```

**Comment by pblottiere** [comment]
@dmarteau 

> Since we have all parameters available internally, may be it could be interesting to have a constructor method for building directly a QgsWmsParameters frow another QgsServerParameters ?

Yes you're right. I'll take a look asap.

**Comment by pblottiere** [comment]
@elpaso What do you think about merging this PR?

**Comment by elpaso** [comment]
@pblottiere @dmarteau  thank you!

**Comment by rldhont** [comment]
What about backporting ? @pblottiere @dmarteau @elpaso

**Pull request merged**